### ISSUE TITLE
Storing and Retrieving test data

### ISSUE THREAD

**Issue by mvdbeek** [galaxyproject/galaxy#7900]
TL;DR: The way we store and retrieve tool (and maybe workflow) test data is not optimal.

The most prevalent approach today is to store the test data next to the tool wrappers in git, and then to upload the tool together with the test data to the tool shed. When installing a repo from the tool shed we also clone the test data.

This is problematic for a couple of reasons:
  - The git and tool shed repository size can become unpractically large
  - We need significant amount of storage for installed tools
  - We're limited by the maximum test data size

There were some ideas and propositions in the past for storing test data:
  - Allow referencing arbitrary URIs for fetching test data
  - Provide test data on CVMFS, downloadable via HTTP at test time
  - Have dedicated test data git repository

There are some considerations to each of them:
  - Arbitrary URI: 
        - How do we ensure the test data is/stays available ?
        - How do we know test data is immutable ?
     
  - CVMFS: How do we get the data into CVMFS?
    - Test data deposition via separate PR ?
    - Deposit test data automatically via scanning existing tool shed repos ?
    - Deposit test data with a bot that scans known galaxy tool github repos for test data ?

  - Git test data repo
    - Would have the same issues as storing test data next to tool IMO

**Comment by martenson** [comment]
@bgruening re: git usage for test data:
We should probably avoid distributing test data in git, I agree there. CVMFS has its own versioning system, that does not come with the strings attached like git repo and we can use it to distribute the data as raw files (compared to repositories).

I am not sure if anybody else than me was suggesting new github repo for test distribution, and I meant it only as a browsable backup.

**Comment by mvdbeek** [comment]
> We should probably avoid distributing test data in git

We distribute via hg atm, which is the worst. I think it's still good to have test data associated with commits. That doesn't necessarily have to be in the same repo though.

Did you know that we can download individual files via hg-web ?
E.g https://toolshed.g2.bx.psu.edu/repos/devteam/bowtie2/raw-file/017aba02828d/test-data/bowtie2-fq1.fa
That means we could in principle add another test data resolver that downloads from the tool shed.

**Comment by martenson** [comment]
> I think it's still good to have test data associated with commits. 

I am with you there. URI and md5sum should be fine though.

**Comment by bgruening** [comment]
Do we all agree that the following is a good first step?

`<param name="metaphlan2_file" uri="https://zenodo.org/api/files/8c2cc766-2b68-45bd-a2d3-391acf9bdb1b/genome_masked.fasta" md5="foobar"/>`

**Comment by martenson** [comment]
@bgruening I do.

**Comment by mvdbeek** [comment]
I would use location like we do for other test definitions. in fact j think it probably works now if we parse these in Galaxy s xnl parser
